internal/layer2/announcer.go

```diff
@@ -131,20 +142,52 @@ func (a *Announce) updateInterfaces() {
 	return
 }
 
-func (a *Announce) spam(name string) {
-	// TODO: should abort if we lose control of the IP mid-spam.
-	start := time.Now()
-	for time.Since(start) < 5*time.Second {
-		if err := a.gratuitous(name); err != nil {
-			a.logger.Log("op", "gratuitousAnnounce", "error", err, "service", name, "msg", "failed to make gratuitous IP announcement")
+func (a *Announce) spamLoop() {
```

Comment:
Way more than that. In the past we were spamming only once when we started to be master. Now we can have some split-brain and continue to be master while someone else is, so we "spam" each time we enter SetBalancer. if there is some instability we have say 10 ForceSync, so we don't want to have 10*5 gratuitous annoncement. With this spam loop you spam for 5 sec, once every 1.1s, so with 10 force sync you maybe will spam for 7 sec but only once every 1.1s

Reply:
K8S is declarative, ie the API only tells you the current state, not the transitions, you know the transitions by keeping a local state but in the case of 'split brain' you miss some transition and that's the problem
We cannot just spam when we become master, so at least we need to always spam

Reply:
We are not using k8s for this, we are using memberlist, right? Are the node events no guaranteed to arrive, even later?

If memberlist is not used, I don't think we hit this problem.

Reply:
The spam race condition already exists today but is harder to hit
1) get 1 node 'offline' more than 40s so it becomes not ready in the API
2) when reconnecting have the node become ready before the speaker can reconnect to the API, the speaker miss the transition and think he was master all along

I started to think about how to keep the fast detection of MemberList by PATCHing some object in the API and make all decision based on what we read back from the API (so we have 1 source of truth), but I don't have something good for now.

Reply:
Quite late here, so sorry for the short answer. But I'm not sure how this answers any of the questions [here][link]. Am I missing something?

[link]: https://github.com/metallb/metallb/pull/595#discussion_r434781491

Reply:
As said on slack I think I should move this commit to another PR, so we can restart fresh.
This added complexity is to try to not spam too much in case of network instability. I've seen 'high end' switches crap out when receiving too much IGMP or multicast on some range, I don't want to break some people network with an ARP storm when I know how to limit this storm

Reply:
As you prefer. But I've already posted here several questions that I'd like you to answer so I can understand better. If you open a new PR, please address them all. It took me a long while to review this several times already :)

As I posted in slack (will quote what I said there to have the discussion in one place), the main motivation underneath is to understand why we are going in this direction with the patch and not something that seems more obvious. I really don't see why we are guessing when to send the arp spam (as the patch does), instead of trying to create a unified view and keep alongside the ips map, a map with the usableNodes, for example, and if the usableNodes map changed, spam (to unify both source of truths and spam only when is needed).

**Specially** why not go towards a "know when to spam" if you don't want to "storm" as you just mentioned, but we are basically spamming "just in case".

Reply:
Here is what we are trying to solve: https://github.com/metallb/metallb/issues/584#issuecomment-613796278
We don't need to know if we are master, we need to know if there was a split brain and someone else was also master, and I haven't found an easy and 100% reliable yet
If it's easier we can have a call (~30min/1h, I'm eastern timezone, ping me on slack)

Reply:
Exactly, that can be detected by usableNodes for this service changing.

I won't have time next week for a call, sorry, I'm super busy. Maybe the week after that. But if the reasoning can be in the open, I think it is better for everyone looking at this. So, unless you really prefer to have a call and can wait a week, I'd continue on github :)

Reply:
@rata can you describe how that would work like I did in https://github.com/metallb/metallb/issues/584#issuecomment-613796278, I willing to implement it, but I don't see it

Reply:
@rata please describe 'in detail' how that would work, I really don't see it.